opencog/rule-engine/forwardchainer/ForwardChainer.cc

```diff
@@ -237,3 +244,479 @@ HandleSeq ForwardChainer::get_chaining_result()
+                                          VIRTUAL_LINK))
+                    and
+                    (not fv.varset.empty());
+
+    return is_valid;
+}
+
+void ForwardChainer::validate(Handle hsource, HandleSeq hfocus_set)
+{
+    if (hsource == Handle::UNDEFINED)
+        throw RuntimeException(TRACE_INFO, "ForwardChainer - Invalid source.");
+   //Any other validation here
+}
+
+/**
+ * Gets all unique atoms of in the implicant list of @param r.
+ *
+ * @param r  A rule object
+ *
+ * @return   An unoderedHandleSet of of all unique atoms in the implicant.
+ */
+UnorderedHandleSet ForwardChainer::get_subatoms(Rule *rule)
+{
+    UnorderedHandleSet output_expanded;
+
+    HandleSeq impl_members = rule->get_implicant_seq();
+    for (Handle h : impl_members) {
+        UnorderedHandleSet hs = get_all_unique_atoms(h);
+        hs.erase(h); //Already tried to unify this.
+        output_expanded.insert(hs.begin(), hs.end());
+    }
+
+    return output_expanded;
+}
+
+/**
+ * Derives new rules from @param hrule by replacing variables
+ * with their groundings.
+ *
+ * @param as             An atomspace where the handles dwell.
+ * @param hrule          A handle to BindLink instance
+ * @param vars           The grounded var list in @param hrule
+ * @param var_groundings The set of groundings to each var in @param vars
+ *
+ * @return A HandleSeq of all possible derived rules
+ */
+HandleSeq ForwardChainer::substitute_rule_part(
+        AtomSpace& as, Handle hrule, const std::set<Handle>& vars,
+        const std::vector<std::map<Handle, Handle>>& var_groundings)
+{
+    std::vector<std::map<Handle, Handle>> filtered_vgmap_list;
+
+    //Filter out variables not listed in vars from var-groundings
+    for (const auto& varg_map : var_groundings) {
+        std::map<Handle, Handle> filtered;
+
+        for (const auto& iv : varg_map) {
+            if (find(vars.begin(), vars.end(), iv.first) != vars.end()) {
+                filtered[iv.first] = iv.second;
+            }
+        }
+
+        filtered_vgmap_list.push_back(filtered);
+    }
+
+    HandleSeq derived_rules;
+    BindLinkPtr blptr = BindLinkCast(hrule);
+    //Substitutor st(&as);
+
+    for (auto& vgmap : filtered_vgmap_list) {
+        Handle himplicand = Substitutor::substitute(blptr->get_implicand(), vgmap);
+        //Create the BindLink/Rule by substituting vars with groundings
+        if (contains_atomtype(himplicand, VARIABLE_NODE)) {
+            Handle himplicant = Substitutor::substitute(blptr->get_body(), vgmap);
+
+            //Assuming himplicant's set of variables are superset for himplicand's,
+            //generate varlist from himplicant.
+            Handle hvarlist = as.add_atom(gen_sub_varlist(
+                    himplicant, LinkCast(hrule)->getOutgoingSet()[0]));
+            Handle hderived_rule = as.add_atom(Handle(createBindLink(HandleSeq {
+                    hvarlist, himplicant, himplicand })));
+            derived_rules.push_back(hderived_rule);
```

Comment:
Sooner or later you need to use the standardized apart version of the rule to make sure the non-var VariableNode will be uniquely generated.